EDU-412: Define Action (#2015)
temporalio/documentation: docs-src/concepts/what-is-an-action.md

---
id: what-is-an-action
title: What is an Action?
sidebar_label: Action
description: An Action is the fundamental pricing unit in Temporal Cloud.
tags:
  - term
  - explanation
---

An Action is the fundamental pricing unit in [Temporal Cloud](/concepts/what-is-temporal-cloud).

**Workflows**

- **Workflow started:** Occurs via client start, client Signal-With-Start, [Continue-As-New](/concepts/what-is-continue-as-new), [Child Workflow](/concepts/what-is-a-child-workflow-execution) start, or [Schedule](/concepts/what-is-a-schedule).
  If a Workflow fails to start, an Action is not recorded.

> - **Workflow reset:** Actions that occurred before a [Reset](/concepts/what-is-a-reset) are counted even if they are no longer visible in [Event History](/concepts/what-is-an-event-history).

Reviewer: I didn't realize "Workflow reset" was a type of action here, because the first bullet primed me to think:

- [context]: list of actions

and this feels like:

- [context]: no listed actions, just an FYI note.

I think my issue might be fixed if all bullets had a period > colon. Then I'm more likely to read as:

- [action]. explanatory text / notes

Author: Excellent point. I'll make that change.

- **Workflow cancellation requested:** Includes requests from a client or from within a Workflow to another Workflow.
- **Timer started:** Includes implicit Timers that are started by a Temporal SDK when timeouts are set, such as `AwaitWithTimeout` in Go or `condition` in TypeScript.

> - **Search Attribute updated:** Occurs after a Workflow has started.

Reviewer: ```suggestion
- **Search Attribute updated:** Occurs after a Workflow has started, when Workflow code calls `UpsertSearchAttributes`.
```

- **Signal sent:** Includes sending a [Signal](/concepts/what-is-a-signal) from a client or from within a [Workflow](/concepts/what-is-a-workflow) to another Workflow.
- **Query received:** [Queries](/concepts/what-is-a-query) aren't recorded in Event History.
  A step such as viewing the stack trace in the Temporal Cloud UI results in a Query.
- **Side Effect recorded:** For a mutable [Side Effect](/concepts/what-is-a-side-effect), an Action occurs only when the value changes.

>   (Some SDKs don't support Side Effects.)

Reviewer: ```suggestion
  (Note that some SDKs don't support Side Effects.)
```

as written, I feel like it's trying to explain what came before, which doesn't make sense

**Activities**

- **Activity attempted:** Each [Activity](/concepts/what-is-an-activity) retry is an attempt and thus an Action.

> - **Activity Heartbeat recorded:** Not every SDK call [Activity Heartbeat](/concepts/what-is-an-activity-heartbeat) counts as an Action.
>   Temporal SDKs throttle Heartbeats.

Reviewer: ```suggestion
- **Activity Heartbeat recorded:** Not every heartbeat call from Activity code counts as an Action—only those that reach the Server.
  Temporal SDKs throttle [Activity Heartbeats](/concepts/what-is-an-activity-heartbeat).
```

---
id: what-is-an-action
title: What is an Action?
sidebar_label: Action
description: An Action is the fundamental pricing unit in Temporal Cloud. Each Action is an Event in the Event History of a Workflow Execution.
tags:
  - term
  - explanation
---

An Action is the fundamental pricing unit in [Temporal Cloud](/concepts/what-is-temporal-cloud).
Each Action is an [Event](/concepts/what-is-an-event) in the [Event History](/concepts/what-is-an-event-history) of a [Workflow Execution](/concepts/what-is-a-workflow-execution).
(Be aware that not all Events are Actions.)

**Workflow Actions**

> - Start Workflow Execution.

Reviewer 2: It can be unclear to users using sessions in Go that we run session handling as an activity

> - Reset Workflow Execution.

Reviewer 2: Probably obvious, but can I get confirmation that the actions that happened before reset are still counted? This is another one that's hard to quantify afterwards because the actions are gone.

- Signal-With-Start Workflow Execution.

> - Continue-As-New Workflow Execution.

Reviewer 2: So is continue as new a double charge with start workflow execution? Maybe to make things clear it should be "Start Workflow Execution - can come from client as start, from client as signal with start, from continue as new, or from child workflow start".

> - Start Child Workflow Execution.

Reviewer 2: This can be confusing to readers to include with "Start Workflow Execution". Don't we just charge for "Start Workflow Execution" regardless of whether it's a child or not?

- Upsert Workflow Search Attributes.

**Activity Actions**

- Schedule Activity Task.

> - Activity Retry. Each retry is counted as a separate Action.

Reviewer 2: I am a bit confused here. So just scheduling an activity that may never run is an action? But it's first attempt is not an action but each attempt after the first is? One might expect for simplicity each activity attempt is an action.

> - Record Marker. A Workflow [Worker](/concepts/what-is-a-worker) executed a [Local Activity](/concepts/what-is-a-local-activity) or a [Side Effect](/concepts/what-is-a-side-effect) or is using [Workflow versioning](/workflows#workflow-versioning).

Reviewer 3: I think having Workflow Worker in the description is a bit too verbose here.

Record Marker Events are associated with a Local Activity Execution and/or a Side Effect - I get that - but how does it have to do with Workflow Versioning?
And if I go to that link... where is that aspect described?
I don't see anything related to Record Markers in the versioning info.

> - Record Marker. A Workflow [Worker](/concepts/what-is-a-worker) executed a [Local Activity](/concepts/what-is-a-local-activity) or a [Side Effect](/concepts/what-is-a-side-effect) or is using [Workflow versioning](/workflows#workflow-versioning).

Reviewer 2: So if I want to save some money on retries, I should use local activities since we only record marker on complete? @Sushisource - I thought there was some logic where we're also counting/recording local activity retries, correct?

> **Signal Actions**

Reviewer 2: Need a section for update.

> - Signal Workflow Execution.

Reviewer 2: To confirm, this is an intentional separate charge from signal with start?

> - Signal External Workflow Execution.

Reviewer 2: Should they really have to care about the genesis of the signal? Can't we just say all signals charge?

**Query Actions**

> - Query Workflow.

Reviewer 2: Are we providing a way to tell users today how many queries they have sent to a workflow? All the other actions I can validate via history, but not this one.

> - Query Workflow.

Reviewer 2: It can be unclear to users that if they do so much as click a stack trace tab in the UI they will be charged. Or are we excluding stack trace queries?

**Timer Actions**

> - Start Timer.

Reviewer 2: May need to clarify for users that things that use timeouts (e.g. `AwaitWithTimeout` in Go, giving a timeout for `condition` in TS, etc), those start a timer internally.
